fdbserver/MoveKeys.actor.cpp

```diff
@@ -373,6 +374,186 @@ ACTOR Future<Optional<UID>> checkReadWrite(Future<ErrorOr<GetShardStateReply>> f
+			return false;
+		}
+	} catch (Error& e) {
+		if (e.code() == error_code_actor_cancelled) {
+			throw e;
+		} else {
+			TraceEvent(SevInfo, "ValidateRangeAssignmentFailed")
+			    .errorUnsuppressed(e)
+			    .detail("Context", context)
+			    .detail("AuditRange", keyServerRange);
+			throw audit_storage_failed();
+		}
+	}
+	return true;
+}
+
+ACTOR Future<bool> validateLocationMetadata(Database occ, KeyRange rangeToCompare, std::string context) {
+	if (rangeToCompare.empty()) {
+		TraceEvent(SevWarn, "ValidateRangeAssignmentEmptyInputRange").detail("AuditRange", rangeToCompare);
+		return true;
+	}
+
+	state std::vector<Future<Void>> actors;
+	state Key rangeToReadBegin = rangeToCompare.begin;
+	state KeyRangeRef rangeToRead;
+	state RangeResult readResultKS;
+	state std::unordered_map<UID, RangeResult> readResultSKs;
+	state RangeResult UIDtoTagMap;
+	state bool anyError = false;
+	state Version readAtVersion;
+	state KeyRange completeRangeKS;
+	state KeyRange skToReadRange;
+	state int iterationCount = 1;
+	state double beginTime = now();
+	state Transaction tr(occ);
+
+	try {
+		loop {
+			try {
+				actors.clear();
+				readResultKS.clear();
+				readResultSKs.clear();
+
+				tr.trState->taskID = TaskPriority::MoveKeys;
+				tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+				tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+				rangeToRead = KeyRangeRef(rangeToReadBegin, rangeToCompare.end);
+
+				// Read keyServers
+				actors.push_back(store(readResultKS,
+				                       krmGetRanges(&tr,
+				                                    keyServersPrefix,
+				                                    rangeToRead,
+				                                    CLIENT_KNOBS->KRM_GET_RANGE_LIMIT,
+				                                    CLIENT_KNOBS->KRM_GET_RANGE_LIMIT_BYTES)));
+				actors.push_back(store(UIDtoTagMap, tr.getRange(serverTagKeys, CLIENT_KNOBS->TOO_MANY)));
+				wait(waitForAll(actors));
+				ASSERT(!UIDtoTagMap.more && UIDtoTagMap.size() < CLIENT_KNOBS->TOO_MANY);
+				completeRangeKS = Standalone(KeyRangeRef(rangeToRead.begin, readResultKS.back().key));
+				TraceEvent(SevVerbose, "ValidateRangeAssignmentKeyServersReadDone")
+				    .detail("Range", completeRangeKS)
+				    .detail("ResultSize", readResultKS.size());
+
+				// Read serverKeys
+				actors.clear();
+				state int i = 0;
+				for (; i < readResultKS.size() - 1; ++i) {
+					std::vector<UID> src;
+					std::vector<UID> dest;
+					UID srcID;
+					UID destID;
+					decodeKeyServersValue(UIDtoTagMap, readResultKS[i].value, src, dest, srcID, destID);
+					std::vector<UID> servers(src.size() + dest.size());
+					std::merge(src.begin(), src.end(), dest.begin(), dest.end(), servers.begin());
+					skToReadRange = Standalone(KeyRangeRef(readResultKS[i].key, readResultKS[i + 1].key));
+					for (auto& ssid : servers) {
+						actors.push_back(store(readResultSKs[ssid],
+						                       krmGetRanges(&tr,
```

Comment:
Is it possible to call `validateRangeAssignment` here?

Reply:
Yes. Nice suggestion.